src/drivers/uavcan/sensors/battery.cpp

```diff
@@ -43,30 +43,102 @@ UavcanBatteryBridge::UavcanBatteryBridge(uavcan::INode &node) :
 	ModuleParams(nullptr),
 	_sub_battery(node),
 	_sub_battery_aux(node),
+	_pub_battery_info(node),
+	_pub_battery_info_aux(node),
 	_warning(battery_status_s::BATTERY_WARNING_NONE),
 	_last_timestamp(0)
 {
 }
 
 int UavcanBatteryBridge::init()
 {
-	int res = _sub_battery.start(BatteryInfoCbBinder(this, &UavcanBatteryBridge::battery_sub_cb));
 
-	if (res < 0) {
-		PX4_ERR("failed to start uavcan sub: %d", res);
-		return res;
-	}
+	int32_t uavcan_sub_bat = 0;
+	param_get(param_find("UAVCAN_SUB_BAT"), &uavcan_sub_bat);
+
+	if (uavcan_sub_bat) {
+		int res = _sub_battery.start(BatteryInfoCbBinder(this, &UavcanBatteryBridge::battery_sub_cb));
+
+		if (res < 0) {
+			PX4_ERR("failed to start uavcan sub: %d", res);
+			return res;
+		}
 
-	res = _sub_battery_aux.start(BatteryInfoAuxCbBinder(this, &UavcanBatteryBridge::battery_aux_sub_cb));
+		res = _sub_battery_aux.start(BatteryInfoAuxCbBinder(this, &UavcanBatteryBridge::battery_aux_sub_cb));
 
-	if (res < 0) {
-		PX4_ERR("failed to start uavcan sub: %d", res);
-		return res;
+		if (res < 0) {
+			PX4_ERR("failed to start uavcan sub: %d", res);
+			return res;
+		}
 	}
 
 	return 0;
 }
 
+void UavcanBatteryBridge::update()
+{
+	// TODO: Decide if we want to do this on UavcanSensorBridgeBase::update()
+	// or have it as a SubscriptionCallbackWorkItem. The latter could allow us
+	// to lock onto the exact same rate as the uORB battery_status topic.
+	battery_status_s battery;
```

Comment:
uavcannode::BatteryInfo already has a BroadcastAnyUpdates function which registers a callback to do this. Why not just modify and use that one instead of reimplementing here?

Reply:
I tried, there are problems with having both uavcan and uavcannode enabled. Also, we need battery_info_aux for individual cell voltages.

Reply:
But I --stole-- borrowed as much as possible from the existing implementation :grimacing: